core/schema/service.go

```diff
@@ -278,3 +289,92 @@ func getIDforSchema(ns, schema, dataUUID string) string {
 	key := fmt.Sprintf("%s-%s-%s", ns, schema, dataUUID)
 	return uuid.NewSHA1(uuid.NameSpaceOID, []byte(key)).String()
 }
+
+func (s *Service) DetectSchemaChange(namespace string, schemaName string, fromVersion string, toVersion string, depth string) (*stencilv1beta1.SchemaChangedEvent, error) {
+	ctx := context.Background()
+	fromVer, toVer, err := s.GetVersions(ctx, namespace, schemaName, fromVersion, toVersion)
+	if err != nil {
+		return nil, err
+	}
+	log.Printf("fromVersion, toVersion - %s, %s\n", fromVersion, toVersion)
+	if depth == "" {
+		depth = strconv.Itoa(-1)
+	}
+	depth64, err := strconv.ParseInt(depth, 10, 32)
+	if err != nil {
+		return nil, fmt.Errorf("invalid depth")
+	}
+	_, fromVerData, err := s.Get(ctx, namespace, schemaName, fromVer)
+	if err != nil {
+		return nil, fmt.Errorf("error getting data for version %d - %s", fromVer, err.Error())
+	}
+	_, toVerData, err := s.Get(ctx, namespace, schemaName, toVer)
+	if err != nil {
+		return nil, fmt.Errorf("error getting data for version %d - %s", toVer, err.Error())
+	}
+	req := &changedetector.ChangeRequest{
+		NamespaceID: namespace,
+		SchemaName:  schemaName,
+		OldData:     fromVerData,
+		NewData:     toVerData,
+		Version:     toVer,
+		Depth:       int32(depth64),
+	}
+	sce, err := s.changeDetectorService.IdentifySchemaChange(ctx, req)
+	if err != nil {
+		return sce, fmt.Errorf("got error while identifying schema change for namespace : %s, schema: %s, version: %d, %s", req.NamespaceID, req.SchemaName, req.Version, err.Error())
+	}
+	if err := s.sendNotification(ctx, sce, req, -1); err != nil {
+		return sce, err
+	}
+	return sce, nil
+}
+
+func (s *Service) GetVersions(ctx context.Context, namespace string, schemaName string, fromVersion string, toVersion string) (int32, int32, error) {
```

Comment:
Looks like no need to expose this method. `getVersions`

And rename it to `validateVersions` as you're passing versions as input as well and the method is parsing & validating it.

Reply:
I exposed this method because it is required in a different package 'schema_test' for writing unit tests. If we did not expose this method, how would we test it?

Reply:
oh okay. Test files can be in same package "schema" or "schema_test", by having "_test" package will block to test helper methods like this.

Also the intention is not to somehow test this helper method, but having correct access level for methods (abstraction) is highest priority. In scenarios where the private methods can't be tested, we need to write unit tests for those public methods which calls the private method.

In golang, this is somehow sorted easily by keeping the same package name, "schema", try it out.